src/miniscript/types/extra_props.rs

```diff
@@ -8,6 +8,90 @@ use MiniscriptKey;
+
+/// Helper struct Whether any satisfaction of this fragment contains any timelocks
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Hash)]
+pub struct TimeLockInfo {
+    // csv with heights
+    pub csv_with_height: bool,
+    /// csv with times
+    pub csv_with_time: bool,
+    /// cltv with heights
+    pub cltv_with_height: bool,
+    /// cltv with times
+    pub cltv_with_time: bool,
+    /// combination of any heightlocks and timelocks
+    pub contains_combination: bool,
+}
+
+impl Default for TimeLockInfo {
+    fn default() -> Self {
+        Self {
+            csv_with_height: false,
+            csv_with_time: false,
+            cltv_with_height: false,
+            cltv_with_time: false,
+            contains_combination: false,
+        }
+    }
+}
+
+impl TimeLockInfo {
+    /// Whether the current contains any possible unspendable
+    /// path
+    pub fn contains_unspendable_path(self) -> bool {
+        self.contains_combination
+    }
+
+    // handy function for combining `and` timelocks
+    // This can be operator overloaded in future
+    pub(crate) fn comb_and_timelocks(a: Self, b: Self) -> Self {
+        Self::combine_thresh_timelocks(2, vec![a, b].into_iter())
+    }
+
+    // handy function for combining `or` timelocks
+    // This can be operator overloaded in future
+    pub(crate) fn comb_or_timelocks(a: Self, b: Self) -> Self {
+        Self::combine_thresh_timelocks(1, vec![a, b].into_iter())
+    }
+
+    pub(crate) fn combine_thresh_timelocks<I>(k: usize, sub_timelocks: I) -> TimeLockInfo
+    where
+        I: IntoIterator<Item = TimeLockInfo>,
+    {
+        let mut timelock_info = TimeLockInfo::default();
+        // timelocks calculation
+        // k >= 2 implies it is possible to have timelocks and heightlocks
+        // merged together
+        // timelocks.csv_with_height and timelocks.csv_with_time can be directly
+        // combined together instead of iterating through all the pairs:
+        // consider the two cases:
+        // 1) Only one subfragment contains timelock and hieghtlock: In this case,
+        // timelocks.contains_combination would already by `true`. and other combinations would
+        // not add any effect.
+        // 2) hieghtlocks and timelocks are distributed across two branches:
+        // and we check if any of the existing branches has correponding lock
+        // that cannot be combined
+        // If k == 1, then this is no combination possible.
+        for sub_timelock in sub_timelocks {
+            if k >= 2 {
+                timelock_info.contains_combination |= (timelock_info.csv_with_height
+                    && sub_timelock.csv_with_time)
+                    || (timelock_info.csv_with_time && sub_timelock.csv_with_height)
+                    || (timelock_info.cltv_with_time && sub_timelock.cltv_with_height)
+                    || (timelock_info.cltv_with_height && sub_timelock.cltv_with_time);
+            }
```

Comment:
You can move this block out of the loop and simplify it:
```
if k >= 2 {
    timelock_info.contains_combination |= timelock_info.csv_with_time && sub_timelock.csv_with_height;
    timelock_info.contains_combination |= timelock_info.cltv_with_time && sub_timelock.cltv_with_height;
}
```
(or however you want to format it)

Reply:
Once it is out of the loop you can replace the loop with `fold`, as
```
let mut ret = sub_timelocks.fold(TimelockInfo::Default, |acc, x| TimelockInfo {
    csv_with_time: acc.csv_with_time | x.csv_with_time,
    csv_with_height: acc.csv_with_height | x.csv_with_height,
    cltv_with_time: acc.cltv_with_time | x.cltv_with_time;
    cltv_with_height: acc.cltv_with_height | x.cltv_with_height;
    contains_combination: acc.contains_combination | acc.contains_combination;
});
if k > 1 {
    ret.contains_combination |= ret.csv_with_time && ret.csv_with_height;
    ret.contains_combination |= ret.cltv_with_time && ret.cltv_with_height;
}
```

Reply:
This is incorrect. As it will set `contains_combination` for the following policy too. 
`thresh(2,pk(),pk(),or(time,height))`

Reply:
Unforetunately, I think it is necessary to check k>2 inside of loop.

Reply:
I think your existing code has the same issue.

Reply:
Ah, no, I see -- the distinction is that only the one child has any timelocks at all. Interesting.

Reply:
Ok, so I leave it up to you whether you want to bother using `fold` instead of a loop. But can you add a comment above the `k >= 2` line saying
```
// If more than one branch may be taken, and some other branch has a requirement
// that conflicts with this one, set `contains_combination`
```
or something

Reply:
Changing to fold